graceful/service.go

```diff
@@ -76,69 +76,99 @@ func (s *service) ListenAndServeTLS(ctx context.Context, proto string, addr stri
 	return s.listenAndServe(ctx, proto, addr, httpServer)
 }
 
-func (s *service) ListenAndServe(ctx context.Context, proto string, addr string, handler http.Handler) error {
+func (s *Service) ListenAndServe(ctx context.Context, proto string, addr string, handler http.Handler) error {
 	httpServer := &http.Server{
 		Addr:    addr,
 		Handler: handler,
 	}
 	return s.listenAndServe(ctx, proto, addr, httpServer)
 }
 
-func (s *service) listenAndServe(ctx context.Context, proto string, addr string, server *http.Server) error {
+func (s *Service) listenAndServe(ctx context.Context, _ string, addr string, server *http.Server) error {
+	log := logger.Get(ctx)
+
 	if s.pidFile != "" {
 		pid := os.Getpid()
 		err := os.WriteFile(s.pidFile, []byte(fmt.Sprintf("%d\n", pid)), 0600)
 		if err != nil {
-			return errgo.Notef(err, "fail to write PID file")
+			return errors.Wrap(ctx, err, "fail to write PID file")
 		}
 	}
 
-	ld, err := s.graceful.Listen(proto, addr)
+	// Use tableflip to handle graceful restart requests
+	upg, err := tableflip.New(tableflip.Options{
+		UpgradeTimeout: s.reloadWaitDuration,
+		PIDFile:        s.pidFile,
+	})
 	if err != nil {
-		return errgo.Notef(err, "fail to get listener")
+		log.Fatalln(err)
 	}
+	s.graceful = upg
+
+	// setup the signal handling
+	go s.setupSignals(ctx)
+
+	// Listen must be called before Ready
+	ln, err := upg.Listen("tcp", addr)
+	if err != nil {
+		upg.Stop()
+		log.Fatalln("Can't listen:", err)
+	}
+
+	defer upg.Stop()
 
 	if server.TLSConfig != nil {
-		ld = tls.NewListener(ld, server.TLSConfig)
+		ln = tls.NewListener(ln, server.TLSConfig)
 	}
 
 	s.httpServer = server
 
-	go s.setupSignals(ctx)
+	go func() {
+		err := server.Serve(ln)
+		if !errors.Is(err, http.ErrServerClosed) {
+			log.Println("HTTP server:", err)
```

Comment:
Like below `Println` → `Info` 

So this error is "normal" right?  It's an info to tells the server has shutdown normally?

Reply:
I have changed this to log.Error.

 I think it is better to log an error as we should be made aware if this fails regularly.